### PR TITLE
Add IL battle & Moon Jump features

### PR DESCRIPTION
**Pull request by Nambo0**
Hi complex :D

### CODE REVIEW

File: src/mods/ilbattle.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+static u32 best_score = 0;
+static constexpr u32 SECOND_FRAMES = 60;
+static constexpr u32 MINUTE_FRAMES = SECOND_FRAMES * 60;
+static constexpr u32 HOUR_FRAMES = MINUTE_FRAMES * 60;
+static constexpr u32 CWIDTH = 12;
+static constexpr u32 CHEIGHT = 16;
+static bool invalid_pause = false;
+static bool buzzer_beater_message = false;
+static u32 buzzer_message_count = 0;
+static u32 battle_length = 0;
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: Static non-`const`/`constexpr` variables should start with `s_`

File: src/mods/ilbattle.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    draw::debug_text(X, Y + 2 * CHEIGHT, color, "%d", best_score);
+}
+
+static u32 score_calc(u32 score) {
+    u32 igt_score = mkb::mode_info.stage_time_frames_remaining * 100 / 60;  // Score from timer
+    u32 goal_bonus = 0;                                                     // Blue goal (no bonus)
+    if (mkb::mode_info.entered_goal_type == 1)
+        goal_bonus = 10000;  // Green goal bonus
+    else if (mkb::mode_info.entered_goal_type == 2)
+        goal_bonus = 20000;  // Red goal bonus
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: Nitpick: use curly braces here

File: src/mods/ilbattle.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    if (mkb::mode_info.stage_time_frames_remaining >
+        mkb::mode_info.stage_time_limit / 2) {  // With time bonus
+        return score + (igt_score + goal_bonus) * 2;
+    } else {
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: Nitpick: unnecessary `else`

File: src/mods/ilbattle.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+static u32 convert_battle_length(u8 battle_length_choice) {
+    if (battle_length_choice == 0)
+        return 5 * 60 * 60;
+    else if (battle_length_choice == 1)
+        return 7 * 60 * 60;
+    else if (battle_length_choice == 2)
+        return 10 * 60 * 60;
+    else
+        return 0;
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: Unnecessary `else`s

File: src/mods/ilbattle.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+void disp() {
+    if (pref::get_il_battle_display()) {
+        if (s_state == IlBattleState::BattleDone) {
+            battle_display(draw::RED);
+            if (buzzer_beater_message) {
+                display_buzzer_beater_message();
+            }
+        } else
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: Brace

File: src/mods/ilbattle.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            track_invalid_pauses();
+            track_final_attempt();
+        } else if (s_state == IlBattleState::BattleDone) {  // DELETE THIS IF THERE'S NOTHING HERE
+                                                            // LATER
+        }
+
+        // THESE ARE FOR TESTING, REMOVE THEM PLEASE NAMBO
+        if (pad::button_pressed(mkb::PAD_BUTTON_Y)) {
+            new_battle();
+        }
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: PLEASE REMOVE NAMBO

File: src/mods/ilbattle.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+static void track_final_attempt() {
+    bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
+    // Save time & end battle if: Goal
+    if (mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT || mkb::sub_mode == mkb::SMD_GAME_GOAL_MAIN) {
+        u32 pre_buzzer_time = best_frames;
+        track_best();
+        if (pre_buzzer_time < best_frames) buzzer_beater_message = true;
+        s_state = IlBattleState::BattleDone;
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: Could you replace the `buzzer_beater_message` bool with a `BattleDoneBuzzerBeater` state that substitutes `BattleDone` if it's a buzzer beater?

**Review by ComplexPlane — COMMENTED** [review]
Aside from my little nitpicks, code looks good to me! I'll probably actually test it myself before merging as well though.

**Comment by ComplexPlane** [comment]
Notes from testing:

* In the menus, rename "IL Battles" -> "IL Battle"?
* What would you think about removing the "Y to reset battle option" and just keeping the Stop Battle menu option? Y is kind of easy to press accidentally and reset vs. stop is kind of confusing.
* Leaving the stage / GAME main mode should stop/hide the battle display if it's over (but keep running if the battle is still ongoing in case of a mismenu?)
* Do you think you could add some help info to the IL battle menu, perhaps with a "Help" orange header? Something like: 'to start a battle, enter a stage, select "start new battle", and retry to begin'?

**Comment by Nambo0** [comment]
For your "Y to reset" comment, I added that because I was finding it annoying to repeatedly open the mod menu & sometimes navigate to the proper spot between each battle, since usually when im battling with friends we're doing lots of stages back to back & it helps to make resetting as efficient as possible. Think maybe it'd work better maybe with a hotkey of "Z" instead of Y, or some button combination?

**Comment by ComplexPlane** [comment]
I see. I think a single button hotkey is fine, but in that case I think the IL battle mod itself should go behind a bool preference so the hotkey can be enabled/disabled to avoid conflicts with e.g. savestates, if that makes sense

**Comment by ComplexPlane** [comment]
Might also makes sense to document the hotkey as "start new battle" over "reset" so it's just obvious it does the same thing as hitting the "start new battle" button in the menu

File: src/mods/ilbattle.cpp [hunk 8]
```diff
@@ -176,22 +175,24 @@ void tick() {
...
     if (pref::get_il_battle_display()) {
-        if (s_state == IlBattleState::BattleDone || s_state == IlBattleState::BattleDoneBuzzer) {
+        if (s_state == IlBattleState::WaitForFirstRetry && mkb::main_mode == mkb::MD_GAME) {
+            battle_display(draw::WHITE);
+        } else if (s_state == IlBattleState::BattleRunning || s_state == IlBattleState::BuzzerBeater) {
+            battle_display(draw::LIGHT_GREEN);
+        } else if ((s_state == IlBattleState::BattleDone || s_state == IlBattleState::BattleDoneBuzzer) && mkb::main_mode == mkb::MD_GAME){
             battle_display(draw::RED);
-            if (s_state == IlBattleState::BattleDoneBuzzer) {
+            if (s_state == IlBattleState::BattleDoneBuzzer && mkb::main_mode == mkb::MD_GAME) {
```
Comment (ComplexPlane) [inline src/mods/ilbattle.cpp]: I think this main mode check is redundant?

**Comment by ComplexPlane** [comment]
Does finishing a battle, leaving the stage, and entering a new stage show the results from the previous battle when the display comes back up? I think the time/score should be zeroed when you first enter a level.

**Comment by ComplexPlane** [comment]
Also, you should transition out of the buzzer beater state into the normal stopped state after leaving a stage so you're not still in that state when next entering a level

**Pull request merged**